### PR TITLE
rfc: audio: sof: hdmi audio suspend/resume flow refine

### PR DESCRIPTION
**Pull request by libinyang**
This is a RFC patch.

This patch tries to fix an i915 power_up/power_down issue. In the old code, audio will not release i915 power when suspend. It will holds the i915 power even the audio is already suspended.

Signed-off-by: Libin Yang <libin.yang@intel.com>

### CODE REVIEW

**Comment by libinyang** [comment]
The patch also has an issue: hda_dsp_suspend() will assume i915 power is on (This may be not always right). The original code has the same issue.

I will refine the patch later.

**Comment by libinyang** [comment]
Maybe we need something like the below code:
```
diff --git a/sound/soc/sof/intel/hda-dsp.c b/sound/soc/sof/intel/hda-dsp.c
index 6b67246..7895c34 100644
--- a/sound/soc/sof/intel/hda-dsp.c
+++ b/sound/soc/sof/intel/hda-dsp.c
@@ -359,11 +359,17 @@ static int hda_resume(struct snd_sof_dev *sdev)

 int hda_dsp_resume(struct snd_sof_dev *sdev)
 {
+       int ret;
+
        /* turn display power on */
        hda_codec_i915_get(sdev);

        /* init hda controller. DSP cores will be powered up during fw boot */
-       return hda_resume(sdev);
+       ret = hda_resume(sdev);
+
+       hda_codec_i915_put(sdev);
+
+       return ret;
 }

 int hda_dsp_runtime_resume(struct snd_sof_dev *sdev)
@@ -383,6 +389,7 @@ int hda_dsp_suspend(struct snd_sof_dev *sdev, int state)
        struct hdac_bus *bus = sof_to_bus(sdev);
        int ret;

+       hda_codec_i915_get(sdev);
        /* stop hda controller and power dsp off */
        ret = hda_suspend(sdev, state);
        if (ret < 0) {
```

**Comment by keyonjie** [comment]
> Maybe we need something like the below code:
> 
> ```
> diff --git a/sound/soc/sof/intel/hda-dsp.c b/sound/soc/sof/intel/hda-dsp.c
> index 6b67246..7895c34 100644
> --- a/sound/soc/sof/intel/hda-dsp.c
> +++ b/sound/soc/sof/intel/hda-dsp.c
> @@ -359,11 +359,17 @@ static int hda_resume(struct snd_sof_dev *sdev)
> 
>  int hda_dsp_resume(struct snd_sof_dev *sdev)
>  {
> +       int ret;
> +
>         /* turn display power on */
>         hda_codec_i915_get(sdev);
> 
>         /* init hda controller. DSP cores will be powered up during fw boot */
> -       return hda_resume(sdev);
> +       ret = hda_resume(sdev);
> +
> +       hda_codec_i915_put(sdev);
> +
> +       return ret;
>  }
> 
>  int hda_dsp_runtime_resume(struct snd_sof_dev *sdev)
> @@ -383,6 +389,7 @@ int hda_dsp_suspend(struct snd_sof_dev *sdev, int state)
>         struct hdac_bus *bus = sof_to_bus(sdev);
>         int ret;
> 
> +       hda_codec_i915_get(sdev);
>         /* stop hda controller and power dsp off */
>         ret = hda_suspend(sdev, state);
>         if (ret < 0) {
> ```

@libinyang as we discussed, just remove hda_dsp_suspend/resume() and redirect them to had_suspend/resume should works.

**Comment by keyonjie** [comment]
@plbossart I think this is good refining to make this i915 PM aligned with Takashi's new design(today's cAVS is broken also I believe).

@libinyang let's fix upstream one for cAVS first, and then adopt SOF change based on new Takashi's for-next.

**Comment by plbossart** [comment]
Agree with @keyonjie this is not SOF specific and we need to review this with Takashi for the cAVS case as well. The timing is bad though, between vacation and merge window, likely something for 2019.

**Comment by libinyang** [comment]
Thanks for the review. @keyonjie @plbossart 

Yes. I will submit a patch directly to alsa mail list after internal review.

**Comment by ranj063** [comment]
@keyonjie @libinyang @plbossart I am bit skeptical about this change. I think the trigger for this change was an issue we're seeing on the topic/rcos-sof-v4.14-rebase branch. If that is the case then we dont have Takashi's latest patches in that branch. So I feel like we're chasing it in the wrong place.

**Comment by libinyang** [comment]
@ranj063 Checked with latest code, I found cros-sof-v4.14-rebase has already merge Takashi's latest patches

**Comment by libinyang** [comment]
> @ranj063 Checked with latest code, I found cros-sof-v4.14-rebase has already merge Takashi's latest patches
@ranj063 
Sorry, I misunderstand which patches are you meaning. You are right. Takashi's patches are not merged. However, this bug seems to be caused by dismatch of get-put. My patch may fix such issues too. Need more test on my patch.

**Comment by ranj063** [comment]
@libinyang ok sure. I'm curious to see the results of the test.
As you know the legacy driver uses the same hdac library and they have not observed any such issue, at least not with 4.14.

**Comment by libinyang** [comment]
> @libinyang ok sure. I'm curious to see the results of the test.
> As you know the legacy driver uses the same hdac library and they have not observed any such issue, at least not with 4.14.

I think maybe the test stress is not enough? Do we have plan to port Takashi's patches? If so, we can directly apply my patches.

**Comment by libinyang** [comment]
I reproduced this issues on my UP2 with latest code yesterday noon and today noon. Each reproduce took about 2 hours now. The root cause is just what guessed before. With my patch, I test with UP2 twice, each time about 12 hours, and the bug can't be reproduced.

However, UP2 can't test suspend/resume when playback. So it needs more tests on this scenario.

**Comment by plbossart** [comment]
@libinyang what patch are you referring to, the code is 10+ days old and it's unclear what you are suggesting as a fix.

**Comment by libinyang** [comment]
> @libinyang what patch are you referring to, the code is 10+ days old and it's unclear what you are suggesting as a fix.

@plbossart 

I mean without my patch, I can reproduce the HDMI audio suspend/resume issue on UP2. And after applying my patch, the issue disappeared. The issue is that with google test suite, the stress test will fail when putting the i915 power.

I do have a little update based on the original patch. Keqiao and I are running the more stress tests on several machines. If the test pasts, I will update the patch on github. And on the same time, I will submit the patch to community.

@keqiaozhang 
Have we filed an issue on github for the HDMI audio suspend/resume?

File: sound/soc/codecs/hdac_hdmi.c [hunk 1]
```diff
@@ -1915,6 +1929,7 @@ static void hdmi_codec_complete(struct device *dev)
 	struct hdac_device *hdev = dev_to_hdac_dev(dev);
 	struct hdac_hdmi_priv *hdmi = hdev_to_hdmi_priv(hdev);
 
+	snd_hdac_display_power(hdev->bus, hdev->addr, true);
```
Comment (ranj063) [inline sound/soc/codecs/hdac_hdmi.c:1932]: @libinyang is this needed if you are already powering off the display power after codec_prepare?

File: sound/soc/codecs/hdac_hdmi.c [hunk 2]
```diff
@@ -1895,7 +1895,20 @@ static int hdmi_codec_prepare(struct device *dev)
...
+	// This is a rfc patch. These comments will removed in the formal patch
+	// if this patch will be merged.
+	// We don't need call pm_runtime_get_sync to wake up all the audio
+	// subsystem just for setting the codec register.
+	// Let's directly power i915 up for codec settings and after that
+	// let's power it off.
+	// If we call pm_runtime_get_sync(&hdev->dev); you will find
+	// i915 power will not be released when suspend. The reason is
+	// described later in the "Power management sequences"
+	// what's more call snd_hdac_display_power(bus, hdev->addr, true);
```
Comment (plbossart) [inline sound/soc/codecs/hdac_hdmi.c:1907]: It wouldn't hurt to explain what the problem is with the "Power management sequences". You are referring to this as if reviewers were familiar with the details.
I honestly could not figure out what the benefits of turning things on/off directly might be over using runtime_pm.

File: sound/soc/codecs/hdac_hdmi.c (hunk 1 above)
Comment (libinyang) [reply sound/soc/codecs/hdac_hdmi.c:1932]: In prepare, I called snd_hdac_display_power(false) to power off i915 at last. So I power on i915 here to make the power status be the same as before.

File: sound/soc/codecs/hdac_hdmi.c (hunk 2 above)
Comment (libinyang) [reply sound/soc/codecs/hdac_hdmi.c:1907]: @plbossart 
OK, I will add the comments in the patch for the details.
And yes, the patch will impact on both SOF and cAVS. I will have a test on cAVS.

**Comment by plbossart** [comment]
closing since this was sorted out directly with Takashi on alsa-devel

**Pull request closed without merging**